scipy/sparse/linalg/eigen/arpack/arpack.py

```diff
@@ -1231,8 +1232,16 @@ def eigs(A, k=6, M=None, sigma=None, which='LM', v0=None,
             import warnings
             warnings.warn('M does not have the same type precision as A. '
                           'This may adversely affect ARPACK convergence')
+
     n = A.shape[0]
 
+    if k >= n - 1:
+        import warnings
+        warnings.warn('k greater than/equal to N - 1 for N * N square matrix. '
+                      'Using scipy.linalg.eig instead.',
+                      SparseEfficiencyWarning)
+        return eig(A.todense(), right=return_eigenvectors)
```

Comment:
This should also check if `M` is given, and if yes, solve the generalized eigenvalue problem. (Add also a test.)

Additionally `A` may be a LinearOperator or a dense matrix, which don't have a `todense` method.